pilot/pkg/model/mesh.go

```diff
@@ -0,0 +1,785 @@
+// Package model provides Mesh which is a platform independent
+// abstraction used by Controllers to synchronize the list of endpoints
+// used by the pilot with those available in the platform specific
+// registry. It implements all the necessary logic that's used for service
+// discovery based on routing rules and endpoint subsets.
+// Typical Usage:
+//
+//   import "istio.io/istio/pilot/model"
+//
+//   type MyPlatformController struct {
+//     Mesh
+//   }
+//   ...
+//   pc := MyPlatformController{model.NewMesh()}
+//   ...
+//   var allEndpoints []*model.ServiceInstances
+//   nativeEndpoints = buildYourPlatformServiceEndpointList()
+//   meshEndpoints := make([]*Endpoint, len(nativeEndpoints))
+//	 for idx, nativeEp := range nativeEndpoints {
+//     // Create mesh endpoint from relevant values of nativeEp
+//	   meshEndpoints[idx] = NewEndpoint(......)
+//   }
+//   pc.Reconcile(meshEndpoints)
+//
+package model
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"net"
+	"reflect"
+	"regexp"
+	"sort"
+	"strconv"
+	"sync"
+
+	xdsapi "github.com/envoyproxy/go-control-plane/api"
+	types "github.com/gogo/protobuf/types"
+	multierror "github.com/hashicorp/go-multierror"
+	route "istio.io/api/routing/v1alpha2"
+	"istio.io/istio/pkg/log"
+)
+
+const (
+	// The internal filter name for attributes that should be used for
+	// constructing Istio attributes on the Endpoint in
+	// <a href='https://goo.gl/mFzjpp'>reverse DNS form</a>.
+	istioConfigFilter = "io.istio.istio-config"
+
+	// Key Istio attribute names for mapping endpoints to subsets.
+	// For details, please see
+	// https://istio.io/docs/reference/config/mixer/attribute-vocabulary.html
+
+	// UID is the attribute name for the mesh unique, platform-specific
+	// unique identifier for the server instance of the destination service. Mesh
+	// uses the label to uqniquely identify the endpoint when it receives updates
+	// from the Controller. Example: kubernetes://my-svc-234443-5sffe.my-namespace
+	// No two instances running anywhere in the mesh can have the same value for
+	// UID.
+	UID destinationAttribute = "destination.uid"
+
+	// SERVICE represents the fully qualified __CANONOCAL__ name of the Istio
+	// service, ex: "my-svc.my-namespace.svc.cluster.local".  Mesh uses
+	// this label to identify endpoints belonging to the same service for XDS queries
+	// involving service names. The attribute value is expected to be mesh unique,
+	// meaning services having this name are semantically understood to be the same
+	// modulo version and other labels, irrespective of the platform under which
+	// they are run, the clusters or availability zones they run in.
+	SERVICE destinationAttribute = "destination.service"
+
+	// NAME represents the short part of the service name, ex: "my-svc".
+	NAME destinationAttribute = "destination.name"
+
+	// NAMESPACE represents the namespace part of the destination service,
+	// example: "my-namespace".
+	NAMESPACE destinationAttribute = "destination.namespace"
```

Comment:
No namespace please.. Model is independent of platforms.

Reply:
I believe we're 'importing' the concept of namespace from k8s into platforms, in the sense of
isolation between sets of applications. It's not k8s namespace, but mesh or istio namespace :-)

Reply:
where will it come from though? The service registry is unlikely to have the namespace in the registry.. Namespace was meant to organize configurations into folders and have ACLs..

Reply:
The adapter between the service registry and the central pilots will need a flag ( or a more complicated mapping ), indicating that the cluster belongs to one namespace. 

We can also say that each cluster is in 'default', or in 'remote' or some other fallback namespace - but still allow a remote cluster ( say consul ) to do more advanced stuff, like use labels or other config to create a similar organization. 

In the 'secure' mode - the VMs in the cluster need to be provisioned with certs anyways, 
and the certs will have spiffe names - with namespace - so that'll be the eventual source
of truth.

Reply:
Hmmm ... so why is it so prominent in https://istio.io/docs/reference/config/mixer/attribute-vocabulary.html @rshriram @costinm do you folks know?

Reply:
I don't know. But look at the routing APIs.. There is no namespace there whatsoever.

Reply:
The spiffe name seems like a fine idea. Also I think this aligns with a bunch of other comments around destination-uid. I suspect none of the current labels are right for what we want here - "the mesh global canonical name". @costin could you get @rshriram @louiscryan @geeknoid to agree on a label name?